Dental implant remote signaling devices!

Winterous opened the pull request:
## About The Pull Request

With that PR nerfing voice analysers and grenade triggers in general, I had the idea of a replacement to the time-honoured tradition of code word activated suicide bombs being something that requires the participation of a second human being, making it less, you know, easy to achieve.

The dental implant surgery, previously only allowing pills to be implanted for easy "oh shit I'm dying" heals, now allows you to insert walkie-talkie sized remote signalers into your teeth! How convenient!

You can also insert anomaly cores, and for some reason they don't even show up with an action button, but there's no reason to other than, I guess, having a treasure trove of valuable items lining your teef?

Includes changes to all the references to implanted pills I could find, removing them from heads, and checking someone's mouth with a torch.

There's no way to remove them without cutting apart a severed head, which is the same for implanted pills, except that obviously you can just, swallow the pills to get rid of them; but this really shouldn't matter, I doubt many people will ever have more than 1 signaler at once. Also is signaler misspelled? Should it be signaller?

## Changelog
:cl:
add: The Dental Implant surgery can now be used to implant Remote Signaling Devices into someone's teeth, enabling hands-free activation of the device by pushing the button with your teeth.
add: The Dental Implant surgery can now be used to implant Anomaly Cores into someone's teeth, enabling hands-free showing off your expensive fillings. See how many you can decorate your smile with!
/:cl:

Winterous committed:
The entire thing!

LemonInTheDark commented:
Ah I see this uses pulsed() and no timer, so it's an instant detonation. Ehhhhh if it requires a second person it won't be done nearly as much.

Winterous commented:
I'm pretty sure pulsed() is just when two assembly items interact, like a timer connected to an igniter. Pretty sure it does the same thing as hitting the signal button.

Winterous commented:
Yeah just took a look, pulsed() just triggers activate() with a 3 second cooldown.

LemonInTheDark commented:
oh excuse me, I'm braindead

Doosteroni commented:
>you can just, swallow the pills to get rid of them 

what if it's some badass espionage suicide pill shit?

Winterous commented:
Well then you probably don't want to remove them until they're used.

necromanceranne commented:
Anomaly grills sounds like the kind of body modification of the future I wanna see.

optimumtact commented:
`enabling hands-free and in-crit activation of the device by pushing the button with your teefs.`

This will never get merged.

Winterous commented:
> This will never get merged.

What? Why? Because of the changelog using teefs instead of teeth?

optimumtact commented:
no, because part of the problem with the voice activated signaller was you could activate it in crit, letting you choose when you blew up and making the suicide bomb a no tradeoff decision.

The whole justification for nerfing it compared to the health sensor is that the health sensor comes with the risk of blowing yourself up if you die to environmental effects.

This completely undermines the other PR, and the fact you made it is confusing in of itself.

Winterous commented:
This is in combination with that one nerfing grenades to PRIME rather than detonate when the signal is received, meaning your prospective victims have a good amount of warning.

Furthermore, it requires the cooperation of someone else to do, so you aren't autonomously producing a suicide bomb, whoever you ask to do it might get you interrogated by sec.

Health sensor will also go off without your input; if you're in hardcrit you can't set this off, so you have to decide to do it BEFORE you become completely helpless.

This is less effective and less autonomous than getting a PAI to do the signaling for you.

Winterous commented:
Anyway, are there perhaps changes you would propose that would make this acceptable? What if you can't do it in crit, like with voice analyzers?

optimumtact commented:
I thought I was clear about what part of it was the issue, specifically the part I highlighted, remove activation in crit and the issue is no longer a problem.

Winterous committed:
Revert "Not in crit, future-proof"

This reverts commit d19ae5243706a4efa478ba7e6b20583e6ff63688.

Winterous committed:
Not in crit, future-proof

Winterous commented:
Done, but turns out you can't swallow implanted pills while in crit anyway, so the change does nothing right now. However I always assumed that was the point of dental pills, so submitting that as an issue to seek advice on whether it needs fixing.

Winterous commented:
https://github.com/tgstation/tgstation/issues/52870

Winterous committed:
Unsecured signalers can't be implanted, and anomaly stuff.

Winterous commented:
Ok I fixed an oversight that allowed you to implant unsecured signalers, which would fail to send the signal, and also added a bunch of stuff related to anomaly cores. You can now show off implanted cores with an action button, giving a slight random visible message, and looking inside someone's mouth gives a different message depending on how many cores there are.

I also tested every aspect of it, and made sure it all works properly. Pretty confident this is 100% complete now.

Rohesie commented on code/game/objects/items/devices/flashlight.dm:
```diff
@@ -158,6 +170,19 @@
+					if(anom_count)
+						var/nice = ""
+						if(anom_count < 3)
+							nice = "Nice."
+						else if(anom_count < 5)
+							nice = "Nice!"
+						else if(anom_count < 33)
+							nice = "NICE!!"
+						else
+							nice = "That's more than humans have teeth, HOLY SHIT!!"
+						to_chat(user, "<span class='notice'>[M] has [anom_count] anomaly core[sig_count > 1 ? "s" : ""] decorating their grin. [nice]</span>")
```
```suggestion
							nice = " Nice."
						else if(anom_count < 5)
							nice = " Nice!"
						else if(anom_count < 33)
							nice = " NICE!!"
						else
							nice = " That's more than humans have teeth, HOLY SHIT!!"
						to_chat(user, "<span class='notice'>[M] has [anom_count] anomaly core[sig_count > 1 ? "s" : ""] decorating their grin.[nice]</span>")
```

Winterous replied:
True that makes a lot more sense.

Rohesie commented on code/modules/surgery/bodyparts/dismemberment.dm:
```diff
@@ -307,6 +307,16 @@
+	for(var/datum/action/item_action/hands_free/activate_signaler/AS in owner.actions)
+		AS.Remove(owner)
+		var/obj/sig = AS.target
+		if(sig)
+			sig.forceMove(src)
```
```suggestion
		sig?.forceMove(src)
```

Winterous replied:
That ? is a tricky little bugger. Mostly seen it in like "is x true? if so 1 else 2" in the form of "var? : 1 || 2" or however it's done.

Rohesie commented on code/modules/surgery/bodyparts/dismemberment.dm:
```diff
@@ -307,6 +307,16 @@
+	for(var/datum/action/item_action/hands_free/showoff/SO in owner.actions)
+		SO.Remove(owner)
+		var/obj/anom = SO.target
+		if(anom)
+			anom.forceMove(src)
```
```suggestion
		anom?.forceMove(src)
```

Rohesie commented on code/modules/surgery/bodyparts/dismemberment.dm:
```diff
@@ -424,6 +434,11 @@
 			P.forceMove(C)
 			AP.Grant(C)
 			break
+	for(var/obj/item/assembly/signaler/S in src)
+		for(var/datum/action/item_action/hands_free/activate_signaler/AS in S.actions)
+			S.forceMove(C)
+			AS.Grant(C)
```
Use more informative var names please.

Winterous replied:
This I was just following the form of the existing one with the pill, thought maybe there was a reason for things like this being 1 or 2 characters. I can run through and change them all to be more verbose.

Rohesie commented on code/modules/surgery/bodyparts/head.dm:
```diff
@@ -117,11 +117,15 @@
-					qdel(AP)
-			I.forceMove(T)
+		else if(istype(I, /obj/item/reagent_containers/pill))
+			for(var/datum/action/item_action/hands_free/activate_pill/AP in I.actions)
+				qdel(AP)
+		else if(istype(I, /obj/item/assembly/signaler))
+			for(var/datum/action/item_action/hands_free/activate_signaler/AS in I.actions)
+				qdel(AS)
+			for(var/datum/action/item_action/hands_free/showoff/SO in I.actions)
+				qdel(SO)
```
Please use more informative var names than `AP`, `AS` and `SO`

Rohesie commented on code/modules/surgery/dental_implant.dm:
```diff
@@ -1,31 +1,49 @@
-	P.target = tool
-	P.Grant(target)	//The pill never actually goes in an inventory slot, so the owner doesn't inherit actions from it
+	if(istype(tool,/obj/item/reagent_containers/pill))
+		var/datum/action/item_action/hands_free/activate_pill/P = new(tool)
+		P.button.name = "Activate [tool.name]"
+		P.target = tool
+		P.Grant(target)	//The pill never actually goes in an inventory slot, so the owner doesn't inherit actions from it
+
+	else if(tool.type == /obj/item/assembly/signaler)
+		var/datum/action/item_action/hands_free/activate_signaler/S = new(tool)
+		S.button.name = "Activate [tool.name]"
+		S.target = tool
+		S.Grant(target)
+
+	else if(istype(tool,/obj/item/assembly/signaler/anomaly))
+		var/datum/action/item_action/hands_free/showoff/S = new(tool)
+		S.button.name = "Show off your [tool.name]"
+		S.target = tool
+		S.Grant(target)
```
* These could stand to be redefined procs instead of bit istype check chains.
* Better variable names please.

Rohesie commented on code/modules/surgery/dental_implant.dm:
```diff
@@ -41,3 +59,24 @@
+/datum/action/item_action/hands_free/activate_signaler
+	name = "Activate Signaling Device"
+
+/datum/action/item_action/hands_free/activate_signaler/Trigger()
+	if(!..())
+		return FALSE
```
```suggestion
/datum/action/item_action/hands_free/activate_signaler/Trigger()
	. = ..()
	if(!.)
		return
```

Winterous replied:
Yeah this stuff with the dots is a little bit gibberish to me, still learning.

Rohesie commented on code/modules/surgery/dental_implant.dm:
```diff
@@ -41,3 +59,24 @@
+	name = "Show Off Implant"
+
+/datum/action/item_action/hands_free/showoff/Trigger()
+	if(!..())
+		return FALSE
```
```suggestion
	. = ..()
	if(!.)
		return
```

Rohesie commented:
I'll be honest, I'm not a fan of invisible surgically-activated implants at all.
They are a cool-sounding gimmick, but gameplay-wise they are not very interesting.

Is any maintainer interested in this?

Winterous committed:
Update code/game/objects/items/devices/flashlight.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Winterous committed:
Update code/modules/surgery/bodyparts/dismemberment.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Winterous committed:
Update code/modules/surgery/bodyparts/dismemberment.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Winterous committed:
Update code/modules/surgery/dental_implant.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Winterous committed:
Update code/modules/surgery/dental_implant.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Winterous commented:
I mostly made this as a less autonomous alternative to in-crit voice analyzer activation, to be honest. But after making it I like the idea even more.

I could add a visible tell, if you're like. Like if you're within a few tiles it's like "X moves their jaw strangely." showing people that they have one. I suppose there should probably be one of those for swallowing a pill too, in movies and stuff people always notice the pill swallow and try to force them to throw it up or something.
